mitodl/ocw-studio
# Set up one of the transcoded video outputs to be downloadable

#### Pre-Flight checklist

#### What are the relevant tickets?
Closes #1283 

#### What's this PR do?
- After creating `VideoFile` objects in response to a MediaConvert webhook, moves the associated S3 object for the 16:9 `VideoFile` to the same path as all the site's other resource content, and updates the `WebsiteContent.file` value for that video resource.

#### How should this be manually tested?
- Use the RC settings for `AWS_ACCOUNT_ID`
- On master branch, place 1 small video file in your gdrive `videos_final` folder for a site and then sync.
- Wait 5 minutes or so to make sure the transcode job finished.
- Using Postman or curl, make a request to http://localhost:8043/api/transcode-jobs/ with the following data, and headers `"Authorization: Bearer <settings.API_BEARER_TOKEN>"` and `"Content-Type: application/json"`:
```
{
  "version": "0",
  "id": "c120fe11-87db-c292-b3e5-1cc90740f6e1",
  "detail-type": "MediaConvert Job State Change",
  "source": "aws.mediaconvert",
  "account": "<AWS_ACCOUNT_ID>",
  "time": "2021-08-05T16:52:33Z",
  "region": "us-east-1",
  "resources": [
    "arn:aws:mediaconvert:us-east-1:<AWS_ACCOUNT_ID>:jobs/<VIDEOJOB_ID>"
  ],
  "detail": {
    "timestamp": 1628172900136,
    "accountId": "<AWS_ACCOUNT_ID>",
    "queue": "arn:aws:mediaconvert:us-east-1:<AWS_ACCOUNT_ID>:queues/Default",
    "jobId": "<VIDEOJOB_ID>",
    "status": "COMPLETE",
    "userMetadata": {},
    "outputGroupDetails": [
      {
        "outputDetails": [
          {
            "outputFilePaths": [
              "s3://<AWS_STORAGE_BUCKET_NAME>/aws_mediaconvert_transcodes/<SITE_NAME>/<DRIVE_FILE_ID>/<VIDEO_FILENAME>.mp4"
            ],
            "durationInMs": 132033,
            "videoDetails": {
              "widthInPx": 1280,
              "heightInPx": 720
            }
          },
          {
            "outputFilePaths": [
              "s3://<AWS_STORAGE_BUCKET_NAME>/aws_mediaconvert_transcodes/<SITE_NAME>/<DRIVE_FILE_ID>/<VIDEO_FILENAME>_360p_16_9.mp4"
            ],
            "durationInMs": 132033,
            "videoDetails": {
              "widthInPx": 640,
              "heightInPx": 360
            }
          },
          {
            "outputFilePaths": [
              "s3://<AWS_STORAGE_BUCKET_NAME>/aws_mediaconvert_transcodes/<SITE_NAME>/<DRIVE_FILE_ID>/<VIDEO_FILENAME>_360p_4_3.mp4"
            ],
            "durationInMs": 132033,
            "videoDetails": {
              "widthInPx": 480,
              "heightInPx": 360
            }
          }
        ],
        "type": "FILE_GROUP"
      }
    ]
  }
}
```
- In django admin, verify that there are 3 video files associated with the video, and all their s3 paths start with `settings.VIDEO_S3_TRANSCODE_PREFIX` (default: `aws_mediaconvert_transcodes`)
- Switch to this PR branch and repeat all the above with another video (don't forget to change the variables in the data example above to match the new video/video-job).  This time, after the POST request, the video file with `_360p_16_9` in its name should have an S3 path starting with `courses/<site_name>/` instead of `aws_mediaconvert_transcodes`
- Run `docker-compose run web python manage.py backpopulate_video_downloads`, after which the first video you imported should now have a modified video file (again with `_360p_16_9` in its name) also with an S3 path starting with `courses/<site_name>/`
- Publish the site (don't worry about the pipeline), check the git repo and make sure the `file` value for that video resource is a working S3 url containing `courses/<site_name>/` in its path.
